### ISSUE TITLE
Mongodb issue

### ISSUE THREAD

**Issue by giviz** [apsillers/lords-of-the-fey#13]
Hey buddy !

I was looking a the project and saw a recent commit, are you getting back to work on it ?

The demo link is no longer working, so I tried to install the game but it seems that there is an issue with mongo db.

```
node server.js
body-parser deprecated bodyParser: use individual json/urlencoded middlewares server.js:77:31
```

```
2018-06-08T20:21:07.094-0400 I NETWORK  [listener] connection accepted from 127.0.0.1:46348 #15 (1 connection now open)
2018-06-08T20:21:07.094-0400 I NETWORK  [listener] connection accepted from 127.0.0.1:46350 #16 (2 connections nowopen)
```

```
> show tables
users
> db.users.find()
{ "_id" : ObjectId("5b1b1a0f197dd7b40324ff3b"), "username" : "hello", "hash" : "sha1$b622db74$1$" }
{ "_id" : ObjectId("5b1b1a0f197dd7b40324ff3c"), "username" : "goodbye", "hash" : "sha1$f2717d97$1$" }
```

The game is just loading without any response, my best guess is that it's stuck with the db but I'm not very familiar with Mongo and nodejs and tried to make it work for a couple of hours without success.

Any idea what can be the issue or how I can debug it ?
Any chance you get that demo running again ?

Thanks !

**Comment by apsillers** [comment]
Thanks for the interest!

I've seen this before, when I also tried to get things going after a long break. As I recall, the issue is that the mongodb driver changed its `find` function in a version change.

As I recall, the fix was to pin the mongodb driver version in package.json instead of permissively allowing later versions. So use

     "mongodb": "1.4.29"

instead of the current

     "mongodb": "^1.4.29"

Then rerun `npm install`. I forget if hat broke anything else, though. If it does, you might just be able to remove upgrade symbols from all the versions, to use the exact versions that I used when I first wrote this a couple years ago.

**Comment by giviz** [comment]
Thank you for the fast response !

I tried to fix all the package versions : 

```
{
  "name": "webnoth",
  "description": "Webnoth",
  "version": "0.0.1",
  "dependencies": {
    "express": "4.x",
    "express-session": "1.9.x",
    "cookies": "0.5.x",
    "body-parser": "1.9.x",
    "cookie-parser": "1.3.x",
    "mongodb": "1.4.29",
    "socket.io": "1.0.x",
    "passport": "0.2.x",
    "passport-local": "1.0.0",
    "passport.socketio": "3.5.x",
    "passport-facebook": "1.0.x",
    "passport-twitter": "1.0.x",
    "passport-google-oauth": "0.2.x",
    "connect-mongo": "0.5.x",
    "password-hash": "1.2.1",
    "hbs": "2.7.x"
  },
  "private": true
}
```

I also had to fix an issue : 
```
giviz@localhost:~/lords-of-the-fey$ node server.js
{ Error: Cannot find module '../build/Release/bson'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:548:15)
    at Function.Module._load (internal/modules/cjs/loader.js:475:25)
    at Module.require (internal/modules/cjs/loader.js:598:17)
    at require (internal/modules/cjs/helpers.js:11:18)
    at Object.<anonymous> (/home/giviz/lords-of-the-fey/node_modules/bson/ext/index.js:15:10)
    at Module._compile (internal/modules/cjs/loader.js:654:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:665:10)
    at Module.load (internal/modules/cjs/loader.js:566:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:506:12)
    at Function.Module._load (internal/modules/cjs/loader.js:498:3) code: 'MODULE_NOT_FOUND' }
js-bson: Failed to load c++ bson extension, using pure JS version
body-parser deprecated bodyParser: use individual json/urlencoded middlewares server.js:77:31
```

By correcting paths in

```
node_modules/bson/ext/index.js
```

From 
```
bson = require('../build/Release/bson');
```

To
```
bson = require('bson');
```

Sadly, that still not work and stay on loading the page with no content displayed on the browser.

Any other easy fix that came to mind that I could try ?

Thanks !

**Comment by apsillers** [comment]
Hmm, it looks like the bson message is actually just a warning, because
mongo uses a fallback module instead.

I just tried a clean npm install, and it seems like my copy (with the
version fix) works, with only changes to config.js.

Have you set up a mongo instance, and configured to the code to use it? The
repo doesn't include a mongo instance: you have to

1) install mongo somewhere (or subscribe to a service -- I've used mLab for
a free 500 MB test instance)
2) add the connection string for that mongo instance to your config.js file
(https://docs.mongodb.com/manual/reference/connection-string/)

Have you done those steps? If so, then I'll need to dig deeper. If not, let
me know how it goes.

**Comment by giviz** [comment]
Yes I did set in config.js

```
"mongoString":"mongodb://@localhost:27017/databaseName",
```

Which is the correct link as I can login with it

```
mongo mongodb://@localhost:27017/databaseName
```
And the users table with the 2 users is there.

When I start the script, I see connection to the mongodb

```
2018-06-09T10:13:41.510-0400 I NETWORK  [listener] connection accepted from 127.0.0.1:48888 #1 (1 connection now open)
2018-06-09T10:13:41.510-0400 I NETWORK  [listener] connection accepted from 127.0.0.1:48890 #2 (2 connections now open)
2018-06-09T10:13:41.525-0400 I COMMAND  [conn1] command databaseName.$cmd command: isMaster { ismaster: 1, $db: "databaseName" } numYields:0 reslen:239 locks:{} protocol:op_query 0ms
2018-06-09T10:13:41.526-0400 I COMMAND  [conn2] command databaseName.$cmd command: isMaster { ismaster: 1, $db: "databaseName" } numYields:0 reslen:239 locks:{} protocol:op_query 0ms
```

But nothing more in the log when I load the page :(

**Comment by giviz** [comment]
I've setup a test mongodb on mLab and created the users collection with the 2 documents, configured the new link in config.js.

Same issue :(

So I assume it's more of a library connection issue than an issue with mongodb itself.

**Comment by giviz** [comment]
Updating the package to "mongodb": "2.0" solved the issue !

**Comment by apsillers** [comment]
I've commit a fix to pin mongo to v1.4.29. With that fix I was able to do a `git clone` and `npm install` and get a working service immediately, on both Windows 10 and Ubuntu 16.04. I didn't need to upgrade to 2.0, which has an incompatible API change to the `find` function (or the cursor API for reading results, possibly), as you've discovered in your other issue.

* What operating system are you trying to deploy on?
* Are you sure the previous version wasn't interfering somehow? Try completely deleting the code (include `node_modules`) and retrying `git clone` and `npm install`. (Also do a `npm remove -g mongo` if you installed it globally.)
* Don't edit the `mongo` module -- the  `Cannot find module '../build/Release/bson'` error message is harmless and only indicates that mongo will use a slower fallback module for BSON.
